dissect/target/filesystems/zip.py

```diff
@@ -0,0 +1,153 @@
+from __future__ import annotations
+
+import logging
+import stat
+import zipfile
+from datetime import datetime, timezone
+from typing import BinaryIO, Optional
+
+from dissect.util.stream import BufferedStream
+
+from dissect.target.exceptions import FileNotFoundError, FilesystemError
+from dissect.target.filesystem import (
+    Filesystem,
+    FilesystemEntry,
+    VirtualDirectory,
+    VirtualFile,
+    VirtualFilesystem,
+)
+from dissect.target.helpers import fsutil
+
+log = logging.getLogger(__name__)
+
+
+class ZipFilesystem(Filesystem):
+    """Filesystem implementation for zip files.
+
+    Python does not have symlink support in the zipfile module, so that's not currently supported.
+    See https://github.com/python/cpython/issues/82102 for more information.
+    """
+
+    __fstype__ = "zip"
+
+    def __init__(
+        self,
+        fh: BinaryIO,
+        base: Optional[str] = None,
+        alt_separator: str = "",
+        case_sensitive: bool = True,
+        *args,
+        **kwargs,
+    ):
+        super().__init__(alt_separator=alt_separator, case_sensitive=case_sensitive, *args, **kwargs)
+
+        fh.seek(0)
+
+        self.zip = zipfile.ZipFile(fh, mode="r")
+        self.base = base or ""
+        self.alt_separator = alt_separator
+
+        self._fs = VirtualFilesystem(alt_separator=alt_separator, case_sensitive=case_sensitive)
+
+        for member in self.zip.infolist():
+            mname = member.filename.strip("/")
+            if not mname.startswith(self.base):
+                continue
+
+            rel_name = fsutil.normpath(mname[len(self.base) :])
+
+            # NOTE: Normally we would check here if the member is a symlink or not
+
+            file_entry = ZipFilesystemEntry(self, rel_name, member)
+            self._fs.map_file_entry(rel_name, file_entry)
+
+    @staticmethod
+    def detect(fh: BinaryIO) -> bool:
+        """Detect a zip file on a given file-like object."""
+        offset = fh.tell()
+        try:
+            fh.seek(0)
+            return zipfile.is_zipfile(fh)
+        except Exception as e:
+            log.warning("Failed to detect zip filesystem", exc_info=e)
+            return False
+        finally:
+            fh.seek(offset)
+
+    def get(self, path: str) -> ZipFilesystemEntry:
+        """Returns a ZipFilesystemEntry object corresponding to the given path."""
+        return self._fs.get(path)
+
+
+class ZipFilesystemEntry(VirtualDirectory, VirtualFile):
```

Comment:
I think it would be better and more robust to have a separate `class ZipFilesystemDirectoryEntry(VirtualDirectoryEntry)` which would store the `ZipInfo` `entry` somewhere and override the `stat()` method.

It would also do away with a lot of the overridden methods in this class.

Reply:
Fair enough, I was mostly looking for feedback for specifically this decision.